File: cli/commands/citations/urls.ts
/**
 * `atlas citations urls` — cited sources as copy-pasteable Markdown, one file
 * per chapter.
 *
 * The crude slice of `task:0026`, shipped ahead of any rendering because the
 * edition-2 authors can paste it into a Google Doc and work from it today.
 *
 * Three properties came from the authors actually trying to use it, and each
 * one is the difference between a list and a usable list:
 *
 *   - **One file per chapter.** A single 3,000-line document is not something
 *     anyone pastes into a Doc and reviews.
 *   - **Deduplicated within a section.** A source cited three times in one
 *     section is one entry there. The repetition is real in the prose and
 *     meaningless in a reference list.
 *   - **Titles, not URLs, as the link text.** "Sutton, 2019" next to a bare URL
 *     tells a reader nothing they did not already know; the resolved title is
 *     the thing that identifies the work.
 *
 * Titles come from the CSL store, so this file gets better as
 * `atlas citations resolve` fills it in. Entries not yet resolved fall back to
 * the raw URL rather than being hidden, so what is missing stays visible.
 */
import type { Chapter } from '../../../src/textbook-loader/index.js';
import {
  extractSectionCitations,
  type CitationInstance,
} from '../../../src/textbook-loader/citations/extract.js';
import type { Store } from '../../../src/textbook-loader/citations/store.js';

export type ChapterFile = { filename: string; markdown: string; citations: number; unique: number };

export type UrlReport = {
  files: ChapterFile[];
  totalCitations: number;
  uniqueSources: number;
  unrecognised: number;
  resolvedTitles: number;
};

/** Zero-padded chapter number, so files sort correctly in a directory listing. */
function pad(n: number): string {
  return String(n).padStart(2, '0');
}

/**
 * The display form of one citation.
 *
 * `Author, Year ([Title](url))` — the title is the link text, matching what a
 * reader needs to recognise the work. Where the store has no title yet, the
 * URL is shown bare so the gap is obvious rather than papered over.
 */
function bullet(c: CitationInstance, store: Store): string {
  const label = c.anchorText.trim() || '(no link text)';
  const where = c.origin === 'footnote' ? ` [fn ${c.footnoteNumber}]` : '';
  if (!c.key) return `- ${label}${where} (no URL)`;

  const entry = store[c.key];
  const title = entry?.resolvedBy !== 'anchor' ? entry?.item?.title : undefined;
  return title
    ? `- ${label}${where} ([${escapeMd(String(title))}](${c.key}))`
    : `- ${label}${where} (${c.key})`;
}

/** Markdown link text cannot contain unescaped brackets. */
function escapeMd(s: string): string {
  return s.replace(/([[\]])/g, '\\$1');
}

/**
 * Deduplicate within one section by canonical URL.
 *
 * The first occurrence wins, so the order follows the prose. Citations with no
 * URL (an unlinked footnote citation) are keyed by their text instead, since
 * they have nothing else to be distinguished by.
 */
function dedupe(citations: CitationInstance[]): CitationInstance[] {
  const seen = new Set<string>();
  const out: CitationInstance[] = [];
  for (const c of citations) {
    const key = c.key ?? `text:${c.anchorText.trim()}`;
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(c);
  }
  return out;
}

/** One chapter's file. Pure: chapter and store in, markdown out. */
export function formatChapterMarkdown(
  chapter: Chapter,
  store: Store,
): { markdown: string; citations: CitationInstance[]; unrecognised: CitationInstance[] } {
  const lines: string[] = [];
  const kept: CitationInstance[] = [];
  const unrecognised: CitationInstance[] = [];

  lines.push(`# Chapter ${chapter.number} — ${chapter.title}: cited sources`);
  lines.push('');

  const body: string[] = [];
  for (const section of chapter.sections) {
    const found = extractSectionCitations(section);
    unrecognised.push(...found.filter((c) => c.kind === 'content-link'));
    const cites = dedupe(found.filter((c) => c.kind === 'citation' || c.kind === 'unlinked'));
    kept.push(...cites);

    body.push(`## ${chapter.number}.${section.number} ${section.title}`);
    body.push('');
    if (cites.length === 0) {
      body.push('_No citations in this section._');
    } else {
      for (const c of cites) body.push(bullet(c, store));
    }
    body.push('');
  }

  const uniqueInChapter = new Set(kept.map((c) => c.key).filter(Boolean)).size;
  const withTitles = kept.filter(
    (c) => c.key && store[c.key] && store[c.key].resolvedBy !== 'anchor',
  ).length;

  lines.push(
    `_${kept.length} citations · ${uniqueInChapter} unique sources · ${withTitles} with a resolved title._`,
  );
  lines.push('');
  lines.push(
    'Read straight out of the Google Doc, so it cannot drift from the prose. Duplicates within a',
  );
  lines.push(
    'section are collapsed. An entry showing a bare URL has not had its title looked up yet.',
  );
  lines.push('');
  lines.push(...body);

  if (unrecognised.length > 0) {
    lines.push('---');
    lines.push('');
    lines.push(`## Links not recognised as citations (${unrecognised.length})`);
    lines.push('');
    lines.push('Hyperlinks whose text is prose rather than an author and year. Anything here that');
    lines.push(
      'should be a citation needs its link text changed in the Doc to "Author, Year" form.',
    );
    lines.push('');
    for (const c of dedupe(unrecognised)) {
      lines.push(
        `- ${chapter.number}.${c.sectionNumber} — "${c.anchorText.trim()}" (${c.key ?? c.rawUrl})`,
      );
    }
    lines.push('');
  }

  return { markdown: lines.join('\n'), citations: kept, unrecognised };
}

/**
 * All chapter files plus a whole-book index.
 *
 * The index answers "what does the Atlas cite", which the per-chapter files
 * cannot — and it is the question a bibliography exists for.
 */
export function formatUrlFiles(chapters: Chapter[], store: Store): UrlReport {
  const files: ChapterFile[] = [];
  const all: CitationInstance[] = [];
  let unrecognised = 0;

  for (const chapter of chapters) {
    const { markdown, citations, unrecognised: un } = formatChapterMarkdown(chapter, store);
    all.push(...citations);
    unrecognised += un.length;
    const slug = chapter.slug || `chapter-${chapter.number}`;
    files.push({
      filename: `chapter-${pad(chapter.number)}-${slug}.md`,
      markdown,
      citations: citations.length,
      unique: new Set(citations.map((c) => c.key).filter(Boolean)).size,
    });
  }

  const byKey = new Map<string, string[]>();
  for (const c of all) {
    if (!c.key) continue;
    const labels = byKey.get(c.key) ?? [];
    const label = c.anchorText.trim();
    if (label && !labels.includes(label)) labels.push(label);
    byKey.set(c.key, labels);
  }
  const keys = [...byKey.keys()].sort();

  const index: string[] = [];
  index.push('# AI Safety Atlas — all cited sources');
  index.push('');
  index.push(`_${keys.length} unique sources across ${chapters.length} chapters._`);
  index.push('');
  index.push(
    'One file per chapter sits beside this one. Where a source is cited under more than one',
  );
  index.push(
    'spelling, every spelling is shown — that is how inconsistent citation text gets found.',
  );
  index.push('');
  let resolvedTitles = 0;
  for (const key of keys) {
    const entry = store[key];
    const title = entry?.resolvedBy !== 'anchor' ? entry?.item?.title : undefined;
    if (title) resolvedTitles++;
    const spellings = byKey.get(key)!.join(' / ') || '(no link text)';
    index.push(
      title ? `- ${spellings} ([${escapeMd(String(title))}](${key}))` : `- ${spellings} (${key})`,
    );
  }
  index.push('');
  files.push({
    filename: 'all-sources.md',
    markdown: index.join('\n'),
    citations: all.length,
    unique: keys.length,
  });

  return {
    files,
    totalCitations: all.length,
    uniqueSources: keys.length,
    unrecognised,
    resolvedTitles,
  };
}

File: python/atlas_citations/commands/urls.py
"""``atlas citations urls`` - cited sources as copy-pasteable Markdown, per chapter.

The crude slice of ``task:0026``, shipped ahead of any rendering because the
edition-2 authors can paste it into a Google Doc and work from it today.

Three properties came from the authors actually trying to use it, and each one is
the difference between a list and a usable list:

- **One file per chapter.** A single 3,000-line document is not something anyone
  pastes into a Doc and reviews.
- **Deduplicated within a section.** A source cited three times in one section is
  one entry there. The repetition is real in the prose and meaningless in a
  reference list.
- **Titles, not URLs, as the link text.** "Sutton, 2019" next to a bare URL tells
  a reader nothing they did not already know; the resolved title is the thing
  that identifies the work.

Titles come from the CSL store, so this output gets better as ``atlas citations
resolve`` fills it in. Entries not yet resolved fall back to the raw URL rather
than being hidden, so what is missing stays visible.

Ported from ``urls.ts`` under ``task:0029``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

from ..scan import Chapter, Citation, Scan, ScanError, read_scan
from ..store import Store
from .extract import read_store

_MD_BRACKETS = re.compile(r"([\[\]])")


@dataclass(frozen=True)
class ChapterFile:
    filename: str
    markdown: str
    citations: int
    unique: int


@dataclass(frozen=True)
class UrlReport:
    files: list[ChapterFile]
    total_citations: int
    unique_sources: int
    unrecognised: int
    resolved_titles: int


def _escape_md(text: str) -> str:
    """Markdown link text cannot contain unescaped brackets."""
    return _MD_BRACKETS.sub(r"\\\1", text)


def _resolved_title(store: Store, key: str | None) -> str | None:
    """The store's title for a key, but only once something resolved it.

    An unresolved entry's title is the anchor text, which is already the label -
    showing it as the link text would say the same thing twice and hide the fact
    that nothing has been looked up.
    """
    if not key:
        return None
    entry = store.get(key)
    if not entry or entry.get("resolvedBy") == "anchor":
        return None
    title = entry.get("item", {}).get("title")
    return str(title) if title else None


def _bullet(c: Citation, store: Store) -> str:
    """``Author, Year ([Title](url))``.

    Where the store has no title yet, the URL is shown bare so the gap is obvious
    rather than papered over.
    """
    label = c.anchor_text.strip() or "(no link text)"
    where = f" [fn {c.footnote_number}]" if c.origin == "footnote" else ""
    if not c.key:
        return f"- {label}{where} (no URL)"
    title = _resolved_title(store, c.key)
    if title:
        return f"- {label}{where} ([{_escape_md(title)}]({c.key}))"
    return f"- {label}{where} ({c.key})"


def _dedupe(citations: list[Citation]) -> list[Citation]:
    """Deduplicate within one section by canonical URL.

    The first occurrence wins, so the order follows the prose. Citations with no
    URL (an unlinked footnote citation) are keyed by their text instead, since
    they have nothing else to be distinguished by.
    """
    seen: set[str] = set()
    out: list[Citation] = []
    for c in citations:
        key = c.key or f"text:{c.anchor_text.strip()}"
        if key in seen:
            continue
        seen.add(key)
        out.append(c)
    return out


def format_chapter_markdown(
    chapter: Chapter, store: Store
) -> tuple[str, list[Citation], list[Citation]]:
    """One chapter's file. Pure: chapter and store in, markdown out."""
    kept: list[Citation] = []
    unrecognised: list[Citation] = []
    body: list[str] = []

    for section in chapter.sections:
        unrecognised.extend(c for c in section.citations if c.kind == "content-link")
        cites = _dedupe([c for c in section.citations if c.kind in ("citation", "unlinked")])
        kept.extend(cites)

        body.append(f"## {chapter.number}.{section.number} {section.title}")
        body.append("")
        if not cites:
            body.append("_No citations in this section._")
        else:
            body.extend(_bullet(c, store) for c in cites)
        body.append("")

    unique_in_chapter = len({c.key for c in kept if c.key})
    with_titles = sum(1 for c in kept if _resolved_title(store, c.key))

    lines = [
        f"# Chapter {chapter.number} - {chapter.title}: cited sources",
        "",
        f"_{len(kept)} citations · {unique_in_chapter} unique sources · "
        f"{with_titles} with a resolved title._",
        "",
        "Read straight out of the Google Doc, so it cannot drift from the prose. Duplicates "
        "within a",
        "section are collapsed. An entry showing a bare URL has not had its title looked up yet.",
        "",
        *body,
    ]

    if unrecognised:
        lines.extend(
            [
                "---",
                "",
                f"## Links not recognised as citations ({len(unrecognised)})",
                "",
                "Hyperlinks whose text is prose rather than an author and year. Anything here that",
                'should be a citation needs its link text changed in the Doc to "Author, Year" '
                "form.",
                "",
            ]
        )
        for c in _dedupe(unrecognised):
            target = c.key or c.raw_url
            lines.append(
                f'- {chapter.number}.{c.section_number} - "{c.anchor_text.strip()}" ({target})'
            )
        lines.append("")

    return "\n".join(lines), kept, unrecognised


def format_url_files(scan: Scan, store: Store) -> UrlReport:
    """All chapter files plus a whole-book index.

    The index answers "what does the Atlas cite", which the per-chapter files
    cannot - and it is the question a bibliography exists for.
    """
    files: list[ChapterFile] = []
    everything: list[Citation] = []
    unrecognised_total = 0

    for chapter in scan.chapters:
        markdown, citations, unrecognised = format_chapter_markdown(chapter, store)
        everything.extend(citations)
        unrecognised_total += len(unrecognised)
        slug = chapter.slug or f"chapter-{chapter.number}"
        files.append(
            ChapterFile(
                filename=f"chapter-{chapter.number:02d}-{slug}.md",
                markdown=markdown,
                citations=len(citations),
                unique=len({c.key for c in citations if c.key}),
            )
        )

    # Spellings are collected from EVERY instance in the scan, not from the
    # per-section display lists. Those are deduplicated by URL, so a source cited
    # twice in one section under two spellings contributes only the first - and
    # the index below promises that "every spelling is shown", which is how
    # inconsistent citation text gets found. Reading the deduplicated lists here
    # made that promise false for precisely the within-section case it exists to
    # catch. (The TypeScript original had this defect; see `audit:0011` F9.)
    by_key: dict[str, list[str]] = {}
    for c in scan.all_citations():
        if not c.key or c.kind not in ("citation", "unlinked"):
            continue
        labels = by_key.setdefault(c.key, [])
        label = c.anchor_text.strip()
        if label and label not in labels:
            labels.append(label)

    keys = sorted(by_key)
    index = [
        "# AI Safety Atlas - all cited sources",
        "",
        f"_{len(keys)} unique sources across {len(scan.chapters)} chapters._",
        "",
        "One file per chapter sits beside this one. Where a source is cited under more than one",
        "spelling, every spelling is shown - that is how inconsistent citation text gets found.",
        "",
    ]
    resolved_titles = 0
    for key in keys:
        title = _resolved_title(store, key)
        if title:
            resolved_titles += 1
        spellings = " / ".join(by_key[key]) or "(no link text)"
        index.append(
            f"- {spellings} ([{_escape_md(title)}]({key}))" if title else f"- {spellings} ({key})"
        )
    index.append("")

    files.append(
        ChapterFile(
            filename="all-sources.md",
            markdown="\n".join(index),
            citations=len(everything),
            unique=len(keys),
        )
    )

    return UrlReport(
        files=files,
        total_citations=len(everything),
        unique_sources=len(keys),
        unrecognised=unrecognised_total,
        resolved_titles=resolved_titles,
    )


def citations_urls(root: Path, out_dir: Path | None = None) -> int:
    try:
        scan = read_scan(root)
    except ScanError as exc:
        print(f"atlas: {exc}")
        return 1

    # Titles come from the store, so this output improves as `resolve` fills it
    # in. A missing store is not an error: the files still build, showing URLs.
    store = read_store(root)
    if not store:
        print("[atlas] no citation store yet - run `atlas citations extract`; showing URLs only")

    report = format_url_files(scan, store)
    dest = out_dir or (root / "data" / "citations" / "chapters")
    dest.mkdir(parents=True, exist_ok=True)
    for f in report.files:
        (dest / f.filename).write_text(f.markdown, encoding="utf-8")

    print(
        f"{report.total_citations} citations · {report.unique_sources} unique sources · "
        f"{report.resolved_titles} with a resolved title · "
        f"{report.unrecognised} links not recognised as citations"
    )
    print(f"wrote {len(report.files)} files to {dest}")
    return 0
